new tutorial

PhilipVinc opened the pull request.

PhilipVinc committed:
new tutorial

PhilipVinc commented:
Yeah, of course. (Do you know his GitHub handle?)
I wanted to finish it before, though.

My plan was to show how to implement by hand the expectation values and gradient of energy...

gcarleo commented:
@PhilipVinc it's jrm874

jrm874 commented on docs/tutorials/variational_intro.ipynb:
```diff
@@ -0,0 +1,1625 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkUAAAGwCAYAAACnyRH2AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAABGBklEQVR4nO3deXxU9b3/8ffMJDPZJ2RPICxhC8oiAoYAKgoVvN66UVcsotSt6lWhVrEqVdviUqvXXX+14q61VatovWV1QdZARBAiYQskJGFLJuskmTm/P5KMREJIMMmZSV7Px+M8ZubM90w+x0Myb8/3e77HYhiGIQAAgG7OanYBAAAA/oBQBAAAIEIRAACAJEIRAACAJEIRAACAJEI ...
```
Here I made a conceptual mistake. This line should be replaced by:

"Consider the loss function $E_\\theta$ to be optimized with respect to the set of variational parameters $\\{\\theta\\}$. The goal is to find the change of parameters $\\delta \\theta$ such that $E_{\\theta + \\delta \\theta}$ is minimized. Taylor expanding $E_{\\theta + \\delta \\theta}$ to first order in $\\delta \\theta$: \n"

jrm874 replied:
To keep correcting the mistake this line should be replaced by:

"E_{\\theta+\\delta\\theta} = E_\\theta + \\sum_k\\frac{\\partial E_\\theta}{\\partial\\theta_k}\\delta\\theta_k + \\mathcal{O}\\big((\\delta \\theta)^2\\big) \n "

jrm874 replied:
To keep correcting the mistake this line should be replaced by:

'' Besides minimizing  $E_{\\theta+\\delta\\theta}$ with respect to $\\delta \\theta$, we also require that the distance travelled in parameter space to be a small value $\\epsilon$. Distance in parameter space can be measured by the Euclidean distance: \n''

jrm874 replied:
To keep correcting the mistake this line should be replaced by:

"Thus, we can combine the notion that we aim at minimizing $E_{\\theta + \\delta \\theta}$ and traveling distance  $d^2 = \epsilon$ to obtain the Lagrangian that we want to minimize with respect to the parameter variations $\\delta \\theta$ ( $\\lambda$ is the Lagrange Multiplier):\n"

jrm874 replied:
To keep correcting the mistake this line should be replaced by:

"L = E_{\\theta + \\delta \\theta} + \\lambda (d^2-\\epsilon)" = E_\\theta + \\sum_k\\left( \\frac{\\partial E_\\theta}{\\partial\\theta_k} \\delta\\theta_k + \\lambda ((\\delta\\theta_k)^2 -\\epsilon)\\right).\n

jrm874 replied:
To keep correcting the mistake this line should be replaced by:

"The minimum of that function with respect to $\\delta \\theta$ is reached when $\\frac{\\partial L}{\\partial (\\delta\\theta_m)} = 0$:\n"

jrm874 replied:
To keep correcting the mistake this line should be replaced by:

"L = E_{\\theta + \\delta \\theta} + \\lambda (d^2-\epsilon) = E_\\theta + \\sum_k \\frac{\\partial E_\\theta}{\\partial\\theta_k} \\delta\\theta_k + \\lambda \\left( \\sum_{kk'} g_{kk'}(\\theta) \\delta \\theta_k \\delta \\theta_{k'} -\epsilon \\right).\n"

jrm874 replied:
To keep correcting the mistake this line should be replaced by:

"The minimum of that function with respect to $\\delta \\theta$ is reached when $\\frac{\\partial L}{\\partial (\\delta\\theta_m)} = 0$:\n"

jrm874 reviewed (changes requested):
The tutorial looks good to me! I am suggesting a couple of stylistic changes (mostly to make a bit more formal what I wrote for the lectures).

I also realized that I made a small conceptual mistake when deriving the update rule for the vanilla gradient descent. I provide line-by-line suggestions to address this mistake. (if the line-by-line commenting is too annoying to implement I can also send you the corrected text for the cell)

Let me know what you think :)

gcarleo committed:
Apply suggestions from code review

Co-authored-by: jrm874 <43007996+jrm874@users.noreply.github.com>

gcarleo commented:
thanks a lot @jrm874 ! would you mind committing directly the changes to notebook? I think you should be already able to push to this PR? @PhilipVinc

gcarleo commented:
also, do I remember correctly that we also had a tutorial on fermions? I think it would be nice if at some point we could add that too!

jrm874 committed:
Corrected version of the tutorial

Fixed the conceptual mistakes in the derivation of the gradient descent update rule.

jrm874 commented:
I have implemented the fixes, but it doesn't allow me to commit directly. It says that I don't have write access to netket

jrm874 commented:
> also, do I remember correctly that we also had a tutorial on fermions? I think it would be nice if at some point we could add that too!

Yes, we did a brief tutorial about fermions for part of the class. Although I think it can be significantly improved. What did you have in mind? A very introductory-level tutorial where we discuss the commutation rules of creation/annihilation operators and the sign bookkeeping of Jordan-Wigner mapping? I could write something along those lines with some examples on spin-polarized fermions and Hubbard model

PhilipVinc commented:
@jrm874 you should be able to write now.

jrm874 commented:
Thanks @PhilipVinc! Just pushed the commit.
